marlowe-playground-client/src/Halogen/Monaco.purs

```diff
@@ -1,22 +1,63 @@
 module Halogen.Monaco where
 
+import Control.Monad.State (class MonadState)
 import Data.Either (Either(..))
+import Data.Enum (class BoundedEnum, class Enum)
+import Data.Generic.Rep (class Generic)
+import Data.Generic.Rep.Bounded (genericBottom, genericTop)
+import Data.Generic.Rep.Enum (genericCardinality, genericFromEnum, genericPred, genericSucc, genericToEnum)
+import Data.Generic.Rep.Eq (genericEq)
+import Data.Generic.Rep.Ord (genericCompare)
 import Data.Lens (view)
 import Data.Maybe (Maybe(..))
 import Data.Traversable (for_, traverse)
 import Effect.Aff.Class (class MonadAff)
 import Effect.Class (class MonadEffect, liftEffect)
-import Halogen (HalogenM, RefLabel)
+import Halogen (HalogenM, RefLabel, get, modify_)
 import Halogen as H
 import Halogen.HTML (HTML, div)
 import Halogen.HTML.Properties (class_, ref)
 import Halogen.Query.EventSource (Emitter(..), Finalizer(..), effectEventSource)
-import Monaco (CodeActionProvider, CompletionItemProvider, DocumentFormattingEditProvider, Editor, IMarker, IMarkerData, IPosition, LanguageExtensionPoint, MonarchLanguage, Theme, TokensProvider)
+import Monaco (CodeActionProvider, CompletionItemProvider, DocumentFormattingEditProvider, Editor, EmacsMode, IMarker, IMarkerData, IPosition, LanguageExtensionPoint, MonarchLanguage, Theme, TokensProvider, VimMode)
 import Monaco as Monaco
-import Prelude (Unit, bind, const, discard, pure, unit, void, ($), (>>=))
+import Prelude (class Bounded, class Eq, class Ord, class Show, Unit, bind, const, discard, pure, unit, void, ($), (>>=))
+
+data KeyBindings
+  = DefaultBindings
+  | Vim
+  | Emacs
+
+derive instance genericKeyBindings :: Generic KeyBindings _
+
+instance showKeyBindings :: Show KeyBindings where
+  show DefaultBindings = "Default"
+  show Vim = "Vim"
+  show Emacs = "Emacs"
+
+instance eqKeyBindings :: Eq KeyBindings where
+  eq = genericEq
+
+instance ordKeyBindings :: Ord KeyBindings where
+  compare = genericCompare
+
+instance enumKeyBindings :: Enum KeyBindings where
+  succ = genericSucc
+  pred = genericPred
+
+instance boundedKeyBindings :: Bounded KeyBindings where
+  bottom = genericBottom
+  top = genericTop
+
+instance boundedEnumKeyBindings :: BoundedEnum KeyBindings where
+  cardinality = genericCardinality
+  toEnum = genericToEnum
+  fromEnum = genericFromEnum
 
 type State
-  = { editor :: Maybe Editor }
+  = { editor :: Maybe Editor
+    , vimMode :: Maybe VimMode
+    , emacsMode :: Maybe EmacsMode
```

Comment:
It's not really the case that you can have both enabled at once, is it?

Reply:
You are correct, it's not however I was storing a `dispose()` function for each. I suppose I could store `m ()` that is by default `pure ()`

Reply:
ok, it made more sense as `Effect Unit` but done.